src/geometry/manifold/charted_space.lean

```diff
@@ -407,7 +407,114 @@ by simp [atlas, charted_space.atlas]
   chart_at H x = local_homeomorph.refl H :=
 by simpa using chart_mem_atlas H x
 
+/-- Same thing as `H × H'`. We introduce it for tecnical reasons: a charted space `M` with model `H`
+is a set of local charts from `M` to `H` covering the space. Every space is registered as a charted
+space over itself, using the only chart `id`, in `charted_space_self`. You can also define a product
+of charted space `M` and `M'` (with model space `H × H'`) by taking the products of the charts. Now,
+on `H × H'`, there are two charted space structures with model space `H × H'` itself, the one coming
+from `charted_space_self`, and the one coming from the product of the two `charted_space_self` on
+each component. They are equal, but not defeq (because the product of `id` and `id` is not defeq to
+`id`), which is bad as we know. This expedient of renaiming `H × H'` solves this problem. -/
+def model_prod (H : Type*) (H' : Type*) := H × H'
+
+section
+local attribute [reducible] model_prod
+
+instance model_prod_inhabited {α β : Type*} [inhabited α] [inhabited β] :
+  inhabited (model_prod α β) :=
+⟨(default α, default β)⟩
+
+instance (H : Type*) [topological_space H] (H' : Type*) [topological_space H'] :
+  topological_space (model_prod H H') :=
+by apply_instance
+
+/- Next lemma shows up often when dealing with derivatives, register it as simp. -/
+@[simp, mfld_simps] lemma model_prod_range_prod_id
+  {H : Type*} {H' : Type*} {α : Type*} (f : H → α) :
+  range (λ (p : model_prod H H'), (f p.1, p.2)) = set.prod (range f) univ :=
+by rw prod_range_univ_eq
+
+end
+
+/-- The product of two charted spaces is naturally a charted space, with the canonical
+construction of the atlas of product maps. -/
+instance prod_charted_space (H : Type*) [topological_space H]
+  (M : Type*) [topological_space M] [charted_space H M]
+  (H' : Type*) [topological_space H']
+  (M' : Type*) [topological_space M'] [charted_space H' M'] :
+  charted_space (model_prod H H') (M × M') :=
+{ atlas            :=
+    {f : (local_homeomorph (M×M') (model_prod H H')) |
+      ∃ g ∈ charted_space.atlas H M, ∃ h ∈ (charted_space.atlas H' M'),
+        f = local_homeomorph.prod g h},
+  chart_at         := (λ x: (M × M'), (charted_space.chart_at H x.1).prod (charted_space.chart_at H' x.2)),
+  mem_chart_source :=
+  begin
+    intro x,
+    simp only [local_homeomorph.prod_to_local_equiv, set.mem_prod, mem_chart_source, and_self, local_equiv.prod_source],
+  end,
```

Comment:
```suggestion
  mem_chart_source := λ x, by simp only with mfld_simps,
```
The `simp only [...] with mfld_simps` idiom is pretty fast, and it is more readable than a `simp only [...]` because it highlights whether you use nontrivial statements or not.

Reply:
I am not too sure about what the paradigm `simp with` is. By `[...]` in `simp only [...] with mfld_simps` you mean the same arguments as in `simp only [...]`? Is it just about adding `with mfld_simps`?